File: baekjoon/dynamic/2011/main.cpp
#include <iostream>
#include <cstdio>
#include <cstring>

using namespace std;
const int DIVIDE = 1000000;

int cache[5001];
string s;
int solve(int n);
int main()
{
    memset(cache, -1, sizeof(cache));
    cin >> s;
    int ret = solve(0)%DIVIDE;
    printf("%d\n", ret);
    return 0;
}

int solve(int n)
{
    if(n == s.size()) {
        return 1;
    }
    
    int& ret = cache[n];
    if(ret != -1) return ret;
    ret = 0;
    if(s[n] > '0') ret += solve(n+1)%DIVIDE;
    if(n+1 < s.size()) {
        int tmp = (s[n]-'0')*10 + s[n+1]-'0';
        if(tmp >= 10 && tmp <= 26) ret += solve(n+2)%DIVIDE;
    }
    return ret;
}

File: baekjoon/dynamic/2011/Main.java
import java.util.*;

class Main {
    static final int DIVIDE = 1000000;
    static String s;
    static int[] cache;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        s = sc.nextLine();
        cache = new int[5001];
        Arrays.fill(cache, -1);
        System.out.println(solve(0)%DIVIDE);
    }
    
    public static int solve(int n) {
        if(n == s.length()) return 1;
        
        if(cache[n] != -1) return cache[n];
        cache[n] = 0;
        
        if(s.charAt(n) > '0') cache[n] += solve(n+1)%DIVIDE;
        if(n+1 < s.length()) {
            int tmp = (s.charAt(n)-'0')*10 + s.charAt(n+1)-'0';
            if(tmp >= 10 && tmp <= 26) cache[n] += solve(n+2)%DIVIDE;
        }
        return cache[n];
    }
}